### PR TITLE
Implement sorting header pragmas

### PR DESCRIPTION
**Pull request by tuleism**

### CODE REVIEW

File: src/NormalizePragmas.hs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+isPragma = isPrefixOf pragmaPrefix
+
+isEmptyLine :: String -> Bool
+isEmptyLine s = null s || all isSpace s
+
+isShebang :: String -> Bool
+isShebang = isPrefixOf shebangPrefix
+
+isHeader :: String -> Bool
+isHeader s = isPragma s || isEmptyLine s || isShebang s
```
Comment (jkarni) [inline src/NormalizePragmas.hs]: Can't comments come before language pragmas? And I guess for now we don't worry about CPP'd pragmas?

Comment (tuleism) [inline src/NormalizePragmas.hs]: Yes, also from @soenkehahn @sol we will do the whole file (not only header), only sort for `LANGUAGE` pragmas and GHC options. I have a new commit for that.

File: test/NormalizePragmasSpec.hs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      normalizePragmas (unlines input) `shouldBe` unlines input
+
+    it "preserves lines that are not language pragmas or ghc options" $ do
+      let input = [
+              "{- First line }-"
+            , "{-# LANGUAGE B #-}"
+            , "{- Here we have"
+            , "a multiple line comment -}"
+            , "{-# LANGUAGE A #-}"
+            , "{-# INCLUDE <fp.h> #-}"
```
Comment (soenkehahn) [inline test/NormalizePragmasSpec.hs:55]: Do multi-line `LANGUAGE` pragmas work?

File: test/NormalizePragmasSpec.hs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            , "{-# LANGUAGE Ab #-}"
+            , "{-# OPTIONS_GHC Abc #-}"
+            , "{-# LANGUAGE A #-}"
+            ]
+          output = [
+              "{-# LANGUAGE A #-}"
+            , "{-# LANGUAGE Ab #-}"
+            , "{-# LANGUAGE B #-}"
+            , "{-# OPTIONS_GHC Abc #-}"
+            ]
```
Comment (soenkehahn) [inline test/NormalizePragmasSpec.hs:22]: I wouldn't care about `OPTIONS_GHC` at all. They are so seldom that I'm perfectly fine with formatting them manually.

File: test/NormalizePragmasSpec.hs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    it "preserves lines that are not language pragmas or ghc options" $ do
+      let input = [
+              "{- First line }-"
+            , "{-# LANGUAGE B #-}"
+            , "{- Here we have"
+            , "a multiple line comment -}"
+            , "{-# LANGUAGE A #-}"
+            , "{-# INCLUDE <fp.h> #-}"
+            ]
+      normalizePragmas (unlines input) `shouldBe` unlines input
```
Comment (soenkehahn) [inline test/NormalizePragmasSpec.hs:57]: Can we have some other kind of lines interspersed with the `LANGUAGE` pragmas in the test input? The most important one is probably empty lines. But also arbitrary characters could be tested.

File: test/NormalizePragmasSpec.hs (hunk 2 above)
Comment (tuleism) [inline test/NormalizePragmasSpec.hs:55]: Currently no

Comment (soenkehahn) [inline test/NormalizePragmasSpec.hs:55]: I think that's fine for now. Do we want to make sure that they are not destroyed? Meaning it's fine if they are not sorted, but they shouldn't get messed up.

Comment (tuleism) [inline test/NormalizePragmasSpec.hs:55]: As discussed, we leave it like this for now, which means multi-line `LANGUAGE` may be destroyed.

File: test/NormalizePragmasSpec.hs (hunk 3 above)
Comment (tuleism) [inline test/NormalizePragmasSpec.hs:22]: In this case I think it doesn't hurt to have it.

File: test/NormalizePragmasSpec.hs (hunk 4 above)
Comment (tuleism) [inline test/NormalizePragmasSpec.hs:57]: I added one more test case for that

**Comment by qoelet** [comment]
LGTM

**Pull request merged**